client/asset/dcr/dcr.go

```diff
@@ -3078,6 +3091,296 @@ func (dcr *ExchangeWallet) EstimateRegistrationTxFee(feeRate uint64) uint64 {
+		Version:     ver,
+		AssetID:     BipID,
+		CoinID:      toCoinID(&txid, 0),
+		SignedTx:    signedTxBytes,
+		UnsignedTx:  unsignedTxBytes,
+		BondScript:  bondScript,
+		RedeemTx:    redeemTx,
+		BondPrivKey: bondPrivKey,
+		BondAcctSig: ecdsa.SignCompact(priv, acctHash[:], true), // ref compressed pubkey
+	}
+	success = true
+
+	return bond, nil
+}
+
+func (dcr *ExchangeWallet) makeBondRefundTxV0(txid *chainhash.Hash, vout uint32, amt uint64,
+	script []byte, privKey []byte, feeRate uint64) ([]byte, error) {
+	lockTime, pkPush, err := dexdcr.ExtractBondDetailsV0(0, script)
+	if err != nil {
+		return nil, err
+	}
+
+	priv := secp256k1.PrivKeyFromBytes(privKey)
+	pk := priv.PubKey().SerializeCompressed()
+	if !bytes.Equal(pk, pkPush) {
+		return nil, fmt.Errorf("incorrect private key to spend the bond output")
+	}
+
+	redeemMsgTx := wire.NewMsgTx()
+	redeemMsgTx.LockTime = lockTime // match locktime types between tx and stack locktime, and >= cltv lock time
+	bondPrevOut := wire.NewOutPoint(txid, vout, wire.TxTreeRegular)
+	txIn := wire.NewTxIn(bondPrevOut, int64(amt), []byte{})
+	txIn.Sequence = wire.MaxTxInSequenceNum - 1 // not finalized, do not disable cltv
+	redeemMsgTx.AddTxIn(txIn)
+
+	// Calculate fees and add the refund output.
+	redeemSize := redeemMsgTx.SerializeSize() + dexdcr.RedeemBondSigScriptSize + dexdcr.P2PKHOutputSize
+	fee := feeRate * uint64(redeemSize)
+	if fee > amt {
+		return nil, fmt.Errorf("irredeemable bond at fee rate %d atoms/byte", feeRate)
+	}
+
+	redeemAddr, err := dcr.wallet.InternalAddress(dcr.ctx, dcr.config().primaryAcct)
+	if err != nil {
+		return nil, fmt.Errorf("error getting new address from the wallet: %w", translateRPCCancelErr(err))
+	}
+	redeemScriptVer, redeemPkScript := redeemAddr.PaymentScript()
+	redeemTxOut := newTxOut(int64(amt-fee), redeemScriptVer, redeemPkScript)
+	redeemMsgTx.AddTxOut(redeemTxOut)
+
+	// CalcSignatureHash and ecdsa.Sign with secp256k1 private key.
+	redeemInSig, err := sign.RawTxInSignature(redeemMsgTx, 0, script, txscript.SigHashAll,
+		privKey, dcrec.STEcdsaSecp256k1)
+	if err != nil {
+		return nil, fmt.Errorf("error creating signature for bond redeem input script '%v': %w", redeemAddr, err)
+	}
+
+	bondRedeemSigScript, err := dexdcr.RefundBondScript(script, redeemInSig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build bond redeem input script: %w", err)
+	}
+	redeemMsgTx.TxIn[0].SignatureScript = bondRedeemSigScript
+
+	redeemTx, err := redeemMsgTx.Bytes()
+	if err != nil {
+		return nil, fmt.Errorf("unable to serialize bond redemption tx: %w", err)
+	}
+	return redeemTx, nil
+}
+
+// RefundBond refunds a bond output to a new wallet address given the redeem
+// script and private key. In general, the bond should have been created with
+// this wallet, but this method makes an effort to find and spend any bond.
+func (dcr *ExchangeWallet) RefundBond(ctx context.Context, ver uint16, coinID, script []byte, privKey []byte) ([]byte, error) {
```

Comment:
Thinking of UX, I think automatic bond renewal should be a default. Core can handle that in a two step process too, but a one step process, i.e. a `RenewBond` method, would have some advantages too. Just something to consider.

Reply:
I agree automatic bond maintenance should be the preferred approach with Core.  I allude to this a bit in the integration PR that follows, but haven't done this yet.  I'm imagining the settings page will let you set either a target tier or a target bond strength, perhaps in units of the bond increment, plus a choice of the asset to use.